File: android/src/com/marpies/ane/facebook/functions/FetchDeferredAppLinkFunction.java
/*
 * Copyright (c) 2018 Marcel Piestansky
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.marpies.ane.facebook.functions;

import android.net.Uri;
import android.text.TextUtils;
import com.adobe.fre.FREContext;
import com.adobe.fre.FREObject;
import com.facebook.FacebookSdk;
import com.facebook.applinks.AppLinkData;
import com.marpies.ane.facebook.data.AIRFacebookEvent;
import com.marpies.ane.facebook.utils.AIR;
import com.marpies.ane.facebook.utils.FREObjectUtils;
import com.marpies.ane.facebook.utils.StringUtils;
import org.json.JSONException;
import org.json.JSONObject;

import java.net.URL;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class FetchDeferredAppLinkFunction extends BaseFunction {

	@Override
	public FREObject call( FREContext context, FREObject[] args ) {
		super.call( context, args );

		final int listenerID = FREObjectUtils.getIntFromFREObject( args[0] );

		AppLinkData.fetchDeferredAppLinkData( AIR.getContext().getActivity(), FacebookSdk.getApplicationId(), new AppLinkData.CompletionHandler() {
			@Override
			public void onDeferredAppLinkDataFetched( AppLinkData appLinkData ) {
				Uri targetUri = (appLinkData != null) ? appLinkData.getTargetUri() : null;
				if( targetUri != null ) {
					AIR.log( "Deferred app link: Target url " + targetUri );
					try {
						JSONObject json = new JSONObject();
						json.put( "targetURL", targetUri.toString() );
						/* Try to get query parameters */
						String query = targetUri.getQuery();
						if( query != null ) {
							json.put( "parameters", parametersFromQuery( query ) );
						}
						json.put( "listenerID", listenerID );
						AIR.dispatchEvent( AIRFacebookEvent.DEFERRED_APP_LINK, json.toString() );
					} catch( JSONException e ) {
						e.printStackTrace();
						AIR.dispatchEvent( AIRFacebookEvent.DEFERRED_APP_LINK, StringUtils.getEventErrorJSON(
								listenerID, "Error creating deferred app link data."
						) );
					}
				} else {
					AIR.log( "No deferred app link data found." );
					AIR.dispatchEvent( AIRFacebookEvent.DEFERRED_APP_LINK,
							String.format( StringUtils.locale, "{ \"listenerID\": %d, \"notFound\": \"true\" }", listenerID )
					);
				}
			}
		} );

		return null;
	}

	/**
	 * Gets query parameters separated by comma.
	 * @param query
	 * @return
	 */
	private String parametersFromQuery( String query ) {
		String[] params = query.split( "&" );
		StringBuilder builder = new StringBuilder();
		boolean isFirst = true;
		for( String param : params ) {
			String[] pair = param.split( "=" );
			if( pair.length != 2 ) continue;

        	/* Add comma after previous value */
			if( !isFirst ) builder.append( "," );
			isFirst = false;

			builder.append( pair[0] ); // name
			builder.append( "," );
			builder.append( pair[1] ); // value
		}
		return builder.toString();
	}

}

File: ios/AIRFacebook/Functions/FetchDeferredAppLinkFunction.m
/*
 * Copyright (c) 2018 Marcel Piestansky
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "FetchDeferredAppLinkFunction.h"
#import <AIRExtHelpers/MPFREObjectUtils.h>
#import "AIRFacebook.h"
#import <AIRExtHelpers/MPStringUtils.h>
#import "AIRFacebookEvent.h"

#import <Bolts/Bolts.h>
#import <FBSDKCoreKit/FBSDKCoreKit.h>

NSDictionary* parametersDictionaryFromQueryString( NSString* queryString ) {
    NSMutableDictionary *md = [NSMutableDictionary dictionary];
    NSArray *queryComponents = [queryString componentsSeparatedByString:@"&"];

    for( NSString *s in queryComponents ) {
        NSArray *pair = [s componentsSeparatedByString:@"="];
        if([pair count] != 2) continue;

        NSString *key = pair[0];
        NSString *value = pair[1];

        md[key] = value;
    }

    return md;
}

NSString* parametersFromQuery(NSString* queryString) {
    NSMutableString * result = [NSMutableString string];
    NSArray *queryComponents = [queryString componentsSeparatedByString:@"&"];
    BOOL isFirst = YES;

    for( NSString *s in queryComponents ) {
        NSArray *pair = [s componentsSeparatedByString:@"="];
        if([pair count] != 2) continue;

        /* Add comma after previous value */
        if( !isFirst ) [result appendString:@","];
        isFirst = NO;

        [result appendFormat:@"%@,", pair[0]]; // name
        [result appendString:pair[1]]; // value
    }

    return result;
}

FREObject fb_fetchDeferredAppLink( FREContext context, void *functionData, uint32_t argc, FREObject *argv ) {
    int listenerID = [MPFREObjectUtils getInt:argv[0]];
    
    [FBSDKAppLinkUtility fetchDeferredAppLink:^(NSURL *url, NSError *error) {
        if (error) {
            [AIRFacebook log:[NSString stringWithFormat:@"Received error while fetching deferred app link %@", error]];
            [AIRFacebook dispatchEvent:DEFERRED_APP_LINK withMessage:[MPStringUtils getEventErrorJSONString:listenerID errorMessage:error.localizedDescription]];
        } else if (url) {
            [AIRFacebook log:[NSString stringWithFormat:@"Received deferred app link %@", url]];
            NSMutableDictionary* result = [NSMutableDictionary dictionary];
            result[@"listenerID"] = @(listenerID);
            result[@"targetURL"] = url.absoluteString;
            NSString* query = url.query;
            if( query ) {
                NSString* urlParams = parametersFromQuery( query );
                [AIRFacebook log:[NSString stringWithFormat:@"Deferred app link parsed URL params %@", urlParams]];
                result[@"parameters"] = urlParams;
            }
            [AIRFacebook log:[NSString stringWithFormat:@"Final result %@", result]];
            /* Get JSON string from the result */
            NSString* jsonString = [MPStringUtils getJSONString:result];
            if( jsonString ) {
                [AIRFacebook dispatchEvent:DEFERRED_APP_LINK withMessage:jsonString];
            } else {
                [AIRFacebook dispatchEvent:DEFERRED_APP_LINK withMessage:[MPStringUtils getEventErrorJSONString:listenerID errorMessage:@"Error creating deferred app link data."]];
            }
        } else {
            [AIRFacebook log:@"No deferred app link data found."];
            [AIRFacebook dispatchEvent:DEFERRED_APP_LINK withMessage:
                    [MPStringUtils getSingleValueJSONString:listenerID key:@"notFound" value:@"true"]
            ];
        }
    }];

    return nil;
}